File: Boop/Boop/bridges/lua/bridge.lua
#!/usr/bin/env lua
local state_json = os.getenv('BOOP_STATE') or '{}'
local state = {}
local ok, res = pcall(function() return assert(load('return '..state_json))() end)
if ok then state = res else state = {} end

local MODULE_EXT = os.getenv('BOOP_MODULE_EXT') or '.lua'
local SCRIPT_DIR = os.getenv('BOOP_SCRIPT_DIR') or ''
local LIB_DIR = os.getenv('BOOP_LIB_DIR') or ''
local REQUIRE_NAME = os.getenv('BOOP_REQUIRE_NAME') or 'boop_require'

local loaded = {}
local function _boop_require(path)
  local p = path
  if p:sub(-#MODULE_EXT) ~= MODULE_EXT then p = p .. MODULE_EXT end
  local file
  if p:sub(1,6) == '@boop/' then
    file = LIB_DIR .. '/' .. p:sub(7)
  else
    file = SCRIPT_DIR .. '/' .. p
  end
  if loaded[file] then return loaded[file] end
  local f = loadfile(file)
  if not f then return nil end
  local mod = f() or {}
  loaded[file] = mod
  return mod
end

_G[REQUIRE_NAME] = _boop_require
if REQUIRE_NAME ~= 'boop_require' then
  _G['boop_require'] = _boop_require
end

local State = {}
State.__index = State
function State:new(data)
  local obj = setmetatable({}, self)
  obj.text = data.text
  obj.fullText = data.fullText
  obj.selection = data.selection
  obj.network = data.network
  obj.inserts = {}
  obj.messages = {}
  return obj
end
function State:post_info(msg) table.insert(self.messages, {type='info', message=msg}) end
function State:post_error(msg) table.insert(self.messages, {type='error', message=msg}) end
function State:insert(val) table.insert(self.inserts, val) end
function State:fetch(url, method, body)
  if not self.network then self:post_error('Network permission required'); return nil end
  local cmd = 'curl -sL '
  if method and method ~= 'GET' then cmd = cmd .. '-X '..method..' ' end
  if body then cmd = cmd .. '--data '..string.format('%q', body)..' ' end
  cmd = cmd .. url
  local f = io.popen(cmd, 'r')
  local data = f:read('*a')
  local ok2 = f:close()
  if ok2 then return data else self:post_error('Failed to fetch'); return nil end
end
local function escape(s) return s:gsub('\\','\\\\'):gsub('"','\\"'):gsub('\n','\\n') end
function State:to_json()
  local parts = {
    '"text":"'..escape(self.text or '')..'"',
    '"fullText":"'..escape(self.fullText or '')..'"',
    '"selection":"'..escape(self.selection or '')..'"',
    '"inserts":['..table.concat((function()
      local arr={} for i,v in ipairs(self.inserts) do arr[i]='"'..escape(v)..'"' end return arr end)(),',')..']',
    '"messages":['..table.concat((function()
      local arr={} for i,v in ipairs(self.messages) do arr[i]='{"type":"'..escape(v.type)..'","message":"'..escape(v.message)..'"}' end return arr end)(),',')..']'
  }
  return '{'..table.concat(parts, ',')..'}'
end

local script = arg[1]
local stateObj = State:new(state)
local user = assert(loadfile(script))
if user then user()(stateObj) end
print(stateObj:to_json())

File: Boop/Boop/bridges/perl/bridge.pl
#!/usr/bin/env perl
use strict;
use warnings;
use JSON::PP;
use Encode;
use File::Spec;

my $MODULE_EXT = $ENV{BOOP_MODULE_EXT} || '.pl';
my $SCRIPT_DIR = $ENV{BOOP_SCRIPT_DIR} || '';
my $LIB_DIR = $ENV{BOOP_LIB_DIR} || '';
my $REQUIRE_NAME = $ENV{BOOP_REQUIRE_NAME} || 'boop_require';

my $state_json = $ENV{BOOP_STATE} || '{}';
my $data = decode_json($state_json);

package State;
sub new {
    my ($class, $d) = @_;
    my $self = {
        text => $d->{text},
        fullText => $d->{fullText},
        selection => $d->{selection},
        network => $d->{network},
        inserts => [],
        messages => []
    };
    bless $self, $class;
    return $self;
}
sub post_info { my ($s,$m)=@_; push @{$s->{messages}}, {type=>'info', message=>$m}; }
sub post_error { my ($s,$m)=@_; push @{$s->{messages}}, {type=>'error', message=>$m}; }
sub insert { my ($s,$v)=@_; push @{$s->{inserts}}, $v; }
sub fetch {
    my ($s,$url,$method,$body)=@_;
    return undef unless $s->{network};
    my @cmd = ('curl','-sL');
    push @cmd,('-X',$method) if defined $method && $method ne 'GET';
    push @cmd,('--data',$body) if defined $body;
    push @cmd,$url;
    my $out = `@cmd`;
    if ($?==0){ return Encode::decode('UTF-8',$out); }
    else { $s->post_error('Failed to fetch'); return undef; }
}
sub to_hash { my $s=shift; return { text=>$s->{text}, fullText=>$s->{fullText}, selection=>$s->{selection}, inserts=>$s->{inserts}, messages=>$s->{messages} }; }
1;

package main;

sub _boop_require {
    my ($path) = @_;
    my $p = $path;
    $p .= $MODULE_EXT unless $p =~ /\Q$MODULE_EXT\E$/;
    my $full;
    if ($p =~ /^\@boop\//) {
        $full = File::Spec->catfile($LIB_DIR, substr($p,6));
    } else {
        $full = File::Spec->catfile($SCRIPT_DIR, $p);
    }
    if (-e $full) {
        do $full;
    } else {
        CORE::require($path);
    }
}
no strict 'refs';
*{$REQUIRE_NAME} = \&_boop_require;
if ($REQUIRE_NAME ne 'boop_require') {
    *boop_require = \&_boop_require;
}
use strict 'refs';
my $script = shift @ARGV;
my $state = State->new($data);
no strict 'refs';
&{$REQUIRE_NAME}("./$script");
use strict 'refs';
State::post_error($state,'No main function') unless defined &main;
main($state) if defined &main;
print encode_json($state->to_hash);
